olp-cpp-sdk-core/include/olp/core/client/CancellationContext.h

```diff
@@ -29,58 +30,62 @@ namespace olp {
 namespace client {
 
 /**
- * @brief Wrapper which manages cancellation state for multiple network
+ * @brief Wrapper which manages cancellation state for any asynchronous
  * operations in a thread safe way.
+ * All public APIs are thread safe.
+ * This class is both movable and copyable.
  */
 class CORE_API CancellationContext {
  public:
-  /**
-  * @brief Execute a given cancellable code block if the operation has not yet
-  been cancelled. Otherwise, execute a custom cancellation function.
-  * @param execute_fn A function to execute if this operation is not yet
-  cancelled. This function should return a CancellationToken which
-  CancellationContext will propogate a cancel request to.
-  * @param cancel_fn A function which will be called if this operation is
-  already cancelled.
-  */
-  inline void ExecuteOrCancelled(
-      const std::function<olp::client::CancellationToken()>& execute_fn,
-      const std::function<void()>& cancel_fn) {
-    std::lock_guard<std::recursive_mutex> lock(mutex_);
-
-    if (is_cancelled_) {
-      cancel_fn();
-      return;
-    }
+  using ExecuteFuncType = std::function<CancellationToken()>;
+  using CancelFuncType = std::function<void()>;
 
-    sub_operation_cancel_token_ = execute_fn();
-  }
+  CancellationContext();
 
   /**
-   * @brief Allows the user to cancel an ongoing operation which may consist of
-   * multiple network calls in a threadsafe way.
+   * @brief Execute a given cancellable code block if the operation has not yet
+   * been cancelled. Otherwise, execute a custom cancellation function.
+   * @param execute_fn A function to execute if this operation is not yet
+   * cancelled. This function should return a CancellationToken which
+   * CancellationContext will propogate a cancel request to.
+   * @param cancel_fn A function which will be called if this operation is
+   * already cancelled.
+   * @deprecated Will be removed once TaskScheduler will be used.
    */
-  inline void CancelOperation() {
-    std::lock_guard<std::recursive_mutex> lock(mutex_);
-    if (is_cancelled_) {
-      return;
-    }
+  void ExecuteOrCancelled(const ExecuteFuncType& execute_fn,
```

Comment:
Add also basic tests for _ExecuteOrCancelled_ - to cover those cases:

1. When _ExecuteOrCancelled_ is called, to check that _execute_fn_ is is also invoked;
2. When _CancellationContext_ was cancelled and _ExecuteOrCancelled_ is called, check that _cancel_fn_ is called and _execute_fn_ is not.

Or we do not need to cover @deprecated methods in tests?

Reply:
Well, I hope I can remove it soon.
I don't think we need to explicitly add tests for deprecated methods.